diag_manager/fms_doubly_linked_list.F90

```diff
@@ -0,0 +1,275 @@
+  end function push_at_back
+
+  function node_constructor () result (nd)
+    type(node), allocatable :: nd
+    allocate(nd)
+    nd%data => null()
+    nd%prev => null()
+    nd%next => null()
+    print * , "In node constructor"
+  end function node_constructor
+
+  function linked_list_constructor () result (ll)
+    type(linked_list), allocatable :: ll
+    allocate(ll)
+    allocate(ll%head)
+    allocate(ll%tail)
+    print *, 'associated(ll%head) :' , associated(ll%head), &
+      ' associated(ll%head) :' , associated(ll%head)
+    ll%head%next => ll%tail
+    ll%tail%prev => ll%head
+    print * , "In list constructor"
+  end function linked_list_constructor
+
+
+  function literator_constructor ( fnd, tnd ) result (litr)
+    type (node), pointer  :: fnd
+    type (node), pointer  :: tnd
+    type (literator), allocatable :: litr
+    !type (literator), pointer :: litr
+    allocate(litr)
+    litr%current => fnd
+    litr%end => tnd
+  end function literator_constructor
+
+  function get_size (this) result (sz)
+    class(linked_list), intent(in out) :: this
+    integer  :: sz
+    sz = this%theSize
+  end function get_size
+
+
+  function is_size_zero (this) result (r)
+    class(linked_list), intent(in out) :: this
+    logical :: r
+    if (this%theSize == 0) then
+      r = .true.
+    else
+      r = .false.
+    end if
+  end function is_size_zero
+
+  function get_forward_literator(this) result (litr)
+    class(linked_list), intent(in) :: this
+    class(literator), ALLOCATABLE :: litr
+    litr = literator( this%head%next, this%tail )
+  end function get_forward_literator
+
+
+  !! Should return true unless the currrent is pointing to the next
+  function literator_has_data( this ) result( r )
+    class(literator), intent(in) :: this
+    logical :: r
+    if( associated (this%current, this%end )) then
+      r = .false.
+    else
+      r = .true.
+    end if
+  end function literator_has_data
+
+  !> Move to the next iff there is a next
+  function literator_next( this ) result( status )
+    class(literator), intent(in out ) :: this
+    integer :: status
+    if(this%has_data() .eqv. .true.) then
+      this%current => this%current%next
+    endif
+    status = 0
+  end function literator_next
+
+
+
+  !> Get the current data object pointed to by the iterator.
+  !! function does not allocate or assign the result if
+  !! the user mistakenly called it without data present.
+  function  literator_data( this ) result( rd )
+    class(literator), intent(in) :: this
+    class(*),  pointer  :: rd
+    if (this%has_data() .eqv. .true.) then
+      rd => this%current%data
+    endif
+  end function literator_data
+
+  !! Iterate over all the nodes, remove them and deallocate the client data
+  !! that the node was holding.
+  subroutine clear_all( this  )
+    class(linked_list), intent(inout) :: this
+    class(node), pointer :: nd
+    class(literator), allocatable :: iter
+    class(*),  pointer  :: pdata
```

Comment:
The function needs a doxygen description.
Variables need to be documented in doxygen style

Reply:
Done.